lib/pgp.ml

```diff
@@ -0,0 +1,602 @@
+        (is_new, header_code - 192)
+      else
+        (is_new, (header_code - 128) / 4)
+
+    let get_old_length_size length_tag =
+      match length_tag with
+      | 0 -> Ok 2
+      | 1 -> Ok 3
+      | 2 -> Ok 5
+      | 3 -> Error "Length size not implemented"
+      | _ -> Error "Bad length size"
+
+    let get_old_length cs header_code =
+      get_old_length_size (header_code mod 4) >>= fun n ->
+      match n with
+      | 2 -> Ok (n, Cstruct.get_uint8 cs 1)
+      | 3 -> Ok (n, Cstruct.BE.get_uint16 cs 1)
+      | 5 -> Ok (n, Int32.to_int (Cstruct.BE.get_uint32 cs 1))
+      | _ -> Error "Bad length size"
+
+    let get_new_length cs =
+      let first_octet = Cstruct.get_uint8 cs 1 in
+      if first_octet < 192 then
+        Ok (2, first_octet)
+      else if first_octet < 224 then
+        let second_octet = Cstruct.get_uint8 cs 2 in
+        let length = 192 + second_octet + (256 * (first_octet - 192)) in
+        Ok (3, length)
+      else if first_octet < 255 then
+        Error "Partial body lengths are not treated"
+      else
+        let length = Cstruct.BE.get_uint32 cs 2 in
+        Ok (6, Int32.to_int length)
+
+    let decode cs =
+      let header_code = Cstruct.get_uint8 cs 0 in
+      get_tag header_code >>= fun (is_new, tag) ->
+      let length_infos =
+        if is_new then
+          get_new_length cs
+        else
+          get_old_length cs header_code
+      in
+      match length_infos with
+      | Error _ -> Error Fatal
+      | Ok (header_length, packet_length) -> (
+        match of_int tag with
+        | Ok packet_type ->
+          Ok (header_length, {packet_type; packet_length; is_new})
+        | Error _ -> Error (Header (header_length + packet_length)) )
+  end
+
+  module Id = struct
+    type t =
+      { name : string
+      ; email : string }
+    [@@deriving ord, eq, show]
+
+    let decode cs =
+      let id = Cstruct.to_string cs in
+      let sep_id = String.split_on_char '<' id in
+      let name = String.concat "<" (List.rev (List.tl (List.rev sep_id))) in
+      let email = List.nth sep_id (List.length sep_id - 1) in
+      {name; email = String.sub email 0 (String.length email - 1)}
+  end
+
+  module Public_key = struct
+    module Public_key_value = struct
+      type t =
+        | Rsa of Rsa.Public.t
+        | Dsa of Dsa.Public.t
+        | Elgamal of Elgamal.Public.t
+      [@@deriving ord, eq, show]
+    end
+
+    type t =
+      { version : int
+      ; creation_time : int32
+      ; validity_period : int option
+      ; algo : Algo.Public.t
+      ; public_key : Public_key_value.t }
+    [@@deriving ord, eq, show]
+
+    let decode_public_key (algo : Algo.Public.t) packet ~version =
+      let offset =
+        (*A public key packet has another header*)
+        match version with
+        | 3 ->
+          Ok 8
+          (*and a version 3 public key packet also contains a validity period*)
+        | 4 -> Ok 6
+        | _ -> Error "Public key packet has a bad version"
```

Comment:
It's minor but, to ease debugging, it could be nice to have the version number in the message: `Error (Printf.sprintf "Unexpected public key packet version: %d" version)`